infrastructure/src/main/java/org/corfudb/infrastructure/RemoteMonitoringService.java

```diff
@@ -421,25 +423,45 @@ private synchronized void runDetectionTasks() {
      * @return Detector task.
      */
     private DetectorTask restoreRedundancyAndMergeSegments(Layout layout) {
-        int segmentsCount = layout.getSegments().size();
+        String localEndpoint = serverContext.getLocalEndpoint();
 
-        if (segmentsCount == 1) {
-            log.debug("No segments to merge. Skipping step.");
-            return DetectorTask.SKIPPED;
-        } else if (!mergeSegmentsTask.isDone()) {
-            log.debug("Merge segments task already in progress. Skipping spawning another task.");
+        // Check that the task is not currently running.
+        if(!mergeSegmentsTask.isDone()){
+            log.trace("Merge segments task already in progress. Skipping spawning another task.");
             return DetectorTask.SKIPPED;
         }
 
-        log.debug("Number of segments present: {}. Spawning task to merge segments.", segmentsCount);
+        // Check that the current node can invoke a restoration action,
+        // also verify that the current node is healed (not present in the unresponsive list).
```

Comment:
This creates a strong coupling between the fault detector "service" and any server. One of the goals of the fault detector is that it should be able to run on only a subset of the nodes and this code breaks that design.

Reply:
In other words, what if localEndpoint is not a LogUnit, or if a LogUnit doesn't run a fault detector, then what would happen?

Reply:
```
One of the goals of the fault detector is that it should be able to run on only a subset of the nodes and this code breaks that design.
```
Looking at the current (master) implementation of the failure detector, it does not look like this goal was ever in mind.
 
- Without a failure detector on every node, we would not be able to aggregate the poll report correctly, run detection and workflows. 

- All the restoration workflows are initiated from the failure detector task, including failing, healing and restoring a node. We do not have a workflow manager that receives hints from the failure detector on what workflows to execute and when; the failure detector does this instead. 

- The healing workflow initiated on the failure detector already heals the current node, so detecting a heal locally, then invoking the workflow non-locally to restore the local node does not even make sense; this design fixes this issue. 

If we would want to decouple failure detection from the actual workflow execution, I think we would have to create an intermediary entity that runs on every node and executes the restoration workflows locally but is able to aggregate/decouple calls from various failure detectors. This would require the change in the FD. The current PR builds on the current working implementation of FD.

Reply:
> it does not look like this goal was ever in mind.

Not quite. There's a gap between design and implementation and the current code shouldn't be used as a reference. The current way the server gets deployed might obscure that but it was always a goal to have different servers run independently and ManagementAgent and Management server are really "one server". 

> Without a failure detector on every node, we would not be able to aggregate the poll report correctly, run detection and workflows.

That's not true. The `ManagementAgent` and `ManagementServer` should be able to run independently as a health monitoring server. The reason it may seem like the system cannot run without having a FD on every server is because `NODE_STATE_REQUEST` is implemented in the `ManagementServer` where it should have really been in the `BaseServer`. Essentially, heart-beating a node is generic functionality that should be serviced by any node (i.e. BaseServer) and not only nodes that have a Fault Detector, which is how it is currently. 

> All the restoration workflows are initiated from the failure detector task, including failing, healing and restoring a node. 

They are initiated via RPC, so there is a client/server separation there. If the design guarantees a management server on every node, then what's the point of having two local entities witin the same JVM communicate over RPC, why not just invoke the methods directly? 

> We do not have a workflow manager that receives hints from the failure detector on what workflows to execute and when; the failure detector does this instead.

Yes we do, its the orchestrator. The ManagementAgent invokes workflows on the orchestrator  via RPC, it can either be local or remote. 

One way to observe the implementation gap I mentioned earlier to ask the question why is the `ManagementAgent` calling the  `ManagementServer` via RPC and **not directly**. The answer is, having a client/server interface between the two entities was by design and that's why it exists, or partially exists for that matter. 

My point is this patch makes an assumption that is not true by design. The fact that the code doesn't completely reflect the design is not a good reason to perpetuate the issue and make the implementation further diverge from the design.

Whether the FD design is sound or not is a separate problem and this is not the place to discuss it since this PR doesn't target the FD design. We should have a meeting to solidify the design, because a PR comment alone is not sufficient.

Reply:
```
My point is this patch makes an assumption that is not true by design. The fact that the code doesn't completely reflect the design is not a good reason to perpetuate the issue and make the implementation further diverge from the design.
```

This argument does not make any sense. You can come up with literally billion of ways of potential implementations that are not currently present in the code. Something is meaningless and completely useless it is implemented, and the new code is implemented based on the existing code, and not the code that could have/should have been implemented.

Reply:
If we would ever want to split the different servers so that they run on different nodes, I think we would have bigger issues, than decoupling S/T with F/D, don't you think? -- It's a completely different project that requires a lot of considerations and refactoring just like striping. 
What you are suggesting sounds like a premature optimization based on the design considerations that are not apparent. If we keep thinking like this, we would not move anywhere. I see a lot of TODOs in the code that go back as early as 2017, and I don't see this ever gets fixed. Maybe instead of considering what could have/should have been implemented, we implement a fully working piece of code based on the existing codebase?

Reply:
> This argument does not make any sense. You can come up with literally billion of ways of potential implementations that are not currently present in the code. Something is meaningless and completely useless unless it is implemented, and the new code is implemented based on the existing code, and not the code that could have/should have been implemented.

I think there are better ways to disagree than exaggerate :)

> You can come up with literally billion of ways of potential implementations that are not currently present in the code

Except that I didn't. I pointed a concrete example on master that supports what I am saying. To reiterate, why is the ManagementAgent invoking RPCs on ManagementServer when one class encapsulates the other. Isn't it plausible that was done on purpose to maintain the client/server model?

> If we would ever want to split the different servers so that they run on different nodes, I think we would have bigger issues

Decoupling the servers is a goal that we should work towards. One of the main core ideas behind Corfu Log is decoupling sequencing from storage units, so I'm not sure why you believe that this decoupling was never in mind. 

> What you are suggesting sounds like a premature optimization based on the design considerations that are not apparent. If we keep thinking like this, we would not move anywhere.

I didn't suggest any optimizations at all and It seems like you have some inaccurate preconceived opinions, but maybe I can correct them here. I have strong bias for action, so I do value pragmatism.

I'm simply saying we have a design for a purpose: we should try to follow it. Now, if the design is refined over time, or if we have technical debt that prevents us from following some parts of the design then it might be reasonable to veer a little. 

Anyways, the purpose of my initial comment was not to pontificate, but rather ponder the question, what is the cost of having any FD send the heal request and not just the local one?